### PR TITLE
SparseArrays: Speed up right-division by diagonal matrices

### PR DESCRIPTION
**Pull request by eschnett**

### CODE REVIEW

File: stdlib/SparseArrays/src/linalg.jl [hunk 1]
```diff
@@ -156,6 +156,14 @@ function (*)(A::AbstractSparseMatrixCSC, D::Diagonal)
     T = Base.promote_op(*, eltype(D), eltype(A))
     mul!(LinearAlgebra.copy_oftype(A, T), A, D)
 end
+function (\)(D::Diagonal, A::AbstractSparseMatrixCSC)
+    T = Base.promote_op(\, eltype(D), eltype(A))
+    ldiv!(LinearAlgebra.copy_oftype(A, T), D, A)
+end
+function (/)(A::AbstractSparseMatrixCSC, D::Diagonal)
+    T = Base.promote_op(/, eltype(D), eltype(A))
+    rdiv!(LinearAlgebra.copy_oftype(A, T), A, D)
```
Comment (dkarrasch) [inline stdlib/SparseArrays/src/linalg.jl:165]: CI suggests there is no such `rdiv!` method...

File: stdlib/SparseArrays/src/linalg.jl [hunk 2]
```diff
@@ -156,6 +156,14 @@ function (*)(A::AbstractSparseMatrixCSC, D::Diagonal)
     T = Base.promote_op(*, eltype(D), eltype(A))
     mul!(LinearAlgebra.copy_oftype(A, T), A, D)
 end
+function (\)(D::Diagonal, A::AbstractSparseMatrixCSC)
+    T = Base.promote_op(\, eltype(D), eltype(A))
+    ldiv!(LinearAlgebra.copy_oftype(A, T), D, A)
+end
+function (/)(A::AbstractSparseMatrixCSC, D::Diagonal)
+    T = Base.promote_op(/, eltype(D), eltype(A))
```
Comment (dkarrasch) [inline stdlib/SparseArrays/src/linalg.jl:164]:
```suggestion
    T = Base.promote_op(/, eltype(A), eltype(D))
```

Comment (eschnett) [reply stdlib/SparseArrays/src/linalg.jl]: Done. Thanks.

File: stdlib/SparseArrays/src/linalg.jl [hunk 3]
```diff
@@ -156,6 +156,14 @@ function (*)(A::AbstractSparseMatrixCSC, D::Diagonal)
     T = Base.promote_op(*, eltype(D), eltype(A))
     mul!(LinearAlgebra.copy_oftype(A, T), A, D)
 end
+function (\)(D::Diagonal, A::AbstractSparseMatrixCSC)
+    T = Base.promote_op(\, eltype(D), eltype(A))
+    ldiv!(LinearAlgebra.copy_oftype(A, T), D, A)
```
Comment (dkarrasch) [inline stdlib/SparseArrays/src/linalg.jl:161]:
```suggestion
    ldiv!(D, LinearAlgebra.copy_oftype(A, T))
```

File: stdlib/SparseArrays/src/linalg.jl [hunk 4]
```diff
@@ -156,6 +156,14 @@ function (*)(A::AbstractSparseMatrixCSC, D::Diagonal)
...
+end
+function (/)(A::AbstractSparseMatrixCSC, D::Diagonal)
+    T = Base.promote_op(/, eltype(A), eltype(D))
+    rdiv!(LinearAlgebra.copy_oftype(A, T), A, D)
```
Comment (dkarrasch) [inline stdlib/SparseArrays/src/linalg.jl:165]:
```suggestion
    rdiv!(LinearAlgebra.copy_oftype(A, T), D)
```

**Review by dkarrasch — CHANGES_REQUESTED** [review]
The 2-arg `*div!`s should work, at least they exist.

File: stdlib/SparseArrays/src/linalg.jl (hunk 3 above)
Comment (eschnett) [reply stdlib/SparseArrays/src/linalg.jl]: Thanks. Done.

**Comment by dkarrasch** [comment]
S**t, I thought I wrote a long comment, but must have forgotten to push the button. To make the long story short, the left-division is currently working nicely, there is no need for an additional method. The right-division is not too bad in that it doesn't fall back to generic dense methods or whatever, but can be improved still
```julia
using SparseArrays, LinearAlgebra, BenchmarkTools, Test

A = sprand(10000,10000,0.1);
D = Diagonal(rand(10000));

function mydiv(A::AbstractMatrix, D::Diagonal)
    rdiv!(typeof(oneunit(eltype(A))/oneunit(eltype(D))).(A), D)
end

@test A/D ≈ mydiv(A, D)
@btime mydiv($A, $D);
@btime $A / $D;
@btime $D \ $A; # for comparison with mydiv
```
My function `mydiv` is meant to be the relaxation of

https://github.com/JuliaLang/julia/blob/ac825137958bf0bd4cfd61b3722622c8880146c0/stdlib/LinearAlgebra/src/diagonal.jl#L486-L487

to `AbstractMatrix`es. I'd suggest to relax that method, remove the newly added methods, and keep the tests, maybe adding further sparsity tests.

**Comment by eschnett** [comment]
@dkarrasch Do you suggest to provide a generic method for `AbstractMatrix` that handles division by diagonal matrices? It seems dangerous to apply a point-wise operation to them since we know nothing about their internal storage. We can't guarantee that the result of a division by a diagonal matrix can be stored in their internal representation. For example, this function would fail if called for `UniformScaling`.

Should we instead relax the existing division operator to include sparse matrices only?

**Comment by dkarrasch** [comment]
> Do you suggest to provide a generic method for AbstractMatrix that handles division by diagonal matrices? 

Yes. Look at this one:

https://github.com/JuliaLang/julia/blob/ac825137958bf0bd4cfd61b3722622c8880146c0/stdlib/LinearAlgebra/src/diagonal.jl#L597-L598

If there is an issue with the underlying `ldiv!` for that specific array type, we'll get a `MethodError`, and I think that's just fine. IIRC, `UniformScaling` is not an `AbstractMatrix`. And this one _is_ called currently for the left-division of a sparse matrix by a `Diagonal`.

**Comment by eschnett** [comment]
@dkarrasch Done.

**Comment by dkarrasch** [comment]
Ok, my sincere apologies for the mess. Now we've got a method ambiguity. Apparently, there is a subtle reason for the different signatures of `ldiv!` and `rdiv!`, for why one can handle `AbstractMatrix` and the other is more specialized. So I hope we should be fine with adding
```julia
function (/)(A::AbstractSparseMatrixCSC, D::Diagonal)
    T = typeof(oneunit(eltype(A))/oneunit(eltype(D)))
    rdiv!(LinearAlgebra.copy_oftype(A, T), D)
end
```
to `SparseArrays` (and leaving the one in `diagonal.jl` unchanged compared to `master`).

Initially, my main point was that `D \ A` works just fine, so we don't need to change anything to allow for that. `A / D` also works, but has a little overhead due to the multiple `adjoints`, so we can improve that.

File: stdlib/SparseArrays/test/sparse.jl [hunk 5]
```diff
@@ -519,6 +519,7 @@ dA = Array(sA)
         @test lmul!(Diagonal(bi), copy(dA)) ≈ ldiv!(Diagonal(b), copy(sA))
         @test lmul!(Diagonal(bi), copy(dA)) ≈ ldiv!(transpose(Diagonal(b)), copy(sA))
         @test lmul!(Diagonal(conj(bi)), copy(dA)) ≈ ldiv!(adjoint(Diagonal(b)), copy(sA))
+        @test Diagonal(b) \ copy(sA) == ldiv!(Diagonal(b), copy(sA))
```
Comment (dkarrasch) [inline stdlib/SparseArrays/test/sparse.jl:522]: No need for the copy, and I'd suggest so check for sparsity.
```suggestion
        @test Diagonal(b) \ sA == ldiv!(Diagonal(b), copy(sA))
        @test issparse(Diagonal(b) \ sA)
```

File: stdlib/SparseArrays/test/sparse.jl [hunk 6]
```diff
@@ -527,6 +528,7 @@ dA = Array(sA)
         @test rmul!(copy(dAt), Diagonal(bi)) ≈ rdiv!(copy(sAt), Diagonal(b))
         @test rmul!(copy(dAt), Diagonal(bi)) ≈ rdiv!(copy(sAt), transpose(Diagonal(b)))
         @test rmul!(copy(dAt), Diagonal(conj(bi))) ≈ rdiv!(copy(sAt), adjoint(Diagonal(b)))
+        @test copy(sAt) / Diagonal(b) == rdiv!(copy(dAt), Diagonal(b))
```
Comment (dkarrasch) [inline stdlib/SparseArrays/test/sparse.jl:531]: same here...

**Review by dkarrasch — CHANGES_REQUESTED** [review]
Sorry again for the back and forth, this time just a trivial comment on the tests, and I hope with my last suggestion we'll be all set.

**Comment by eschnett** [comment]
@dkarrasch Done. Thanks for the suggestions; this is now a much better implementation.

**Comment by dkarrasch** [comment]
@eschnett Ok, the tests look good now, but you still have the method ambiguity due to my own bold suggestion (sorry again). You'll need to undo the change in `LinearAlgebra/diagonal.jl`, and add https://github.com/JuliaLang/julia/pull/35533#issuecomment-618254702 to `SparseArrays/linalg.jl`. Then I'd expect this to pass tests.

**Comment by eschnett** [comment]
Sorry, did not see your final comment earlier.

I didn't spot the method ambiguity when I ran the test suite locally.

File: stdlib/LinearAlgebra/src/diagonal.jl [hunk 7]
```diff
@@ -483,8 +483,8 @@ rdiv!(A::AbstractMatrix{T}, adjD::Adjoint{<:Any,<:Diagonal{T}}) where {T} =
 rdiv!(A::AbstractMatrix{T}, transD::Transpose{<:Any,<:Diagonal{T}}) where {T} =
     (D = transD.parent; rdiv!(A, D))
 
-(/)(A::Union{StridedMatrix, AbstractTriangular}, D::Diagonal) =
-    rdiv!((typeof(oneunit(eltype(D))/oneunit(eltype(A)))).(A), D)
+(/)(A::Union{StridedMatrix, AbstractTriangular}, D::Diagonal) = 
+    rdiv!((typeof(oneunit(eltype(D))/oneunit(eltype(A)))).(A), D)
```
Comment (dkarrasch) [inline stdlib/LinearAlgebra/src/diagonal.jl:487]: whitespace
```suggestion
(/)(A::Union{StridedMatrix, AbstractTriangular}, D::Diagonal) =
    rdiv!((typeof(oneunit(eltype(D))/oneunit(eltype(A)))).(A), D)
```

**Pull request merged**